### ISSUE TITLE
Our test result reporting does not take into account fails after test done

### ISSUE THREAD

**Issue by vjrantal** [thaliproject/Thali_CordovaPlugin#691]
Currently, then you run the full coordinated run on desktop, you might see the end report suggesting that everything passed, but in the logs, there might be output like:
```
DEBUG createPeerListener: error on incoming stream - Error: Channel destroyed

ok 331 response body should match testData

ok 332 must be started

not ok 333 .end() called twice
  ---

    operator: fail

    at: end Thali_CordovaPlugin/test/www/jxcore/bv_tests/testThaliMobileNativeWrapper.js:157:35)

  ...
```
The tape test report reports these as failures, but we don't, because we report to coordination server after each test if there were failures, but don't take into account in case a failure in test occurs after the test itself has completed.

This is not super critical and often the issue is just not cleaning up all events handlers at the end of each test. However, this results into tape test report suggest failures we our coordination server thinks all went fine.

**Comment by vjrantal** [comment]
@yaronyg: From my perspective, this can be left to icebox, but wanted to create this issue so that you and others become aware of it.

**Comment by yaronyg** [comment]
Yeah, I saw this yesterday when I was running tests. I actually found and fixed the tests but this particular bug would be a real pain to fix since it requires us to go run back to the coordinate and undo a success. Certainly something for the icebox.

**Comment by yaronyg** [comment]
@artemjackson This is a relatively easy first bug for you. The issue is that when we run a test in node there are three parts to each and every test. There is setup, the test and then tear down. When we run a 'coordinated' test our coordinated test framework which lives under test/TestServer waits for each device to connect to it and say "I'm ready to run setup". So the test server just waits until everyone says "I'm reading to run setup" and then once it gets o.k. from everyone it's been told to expect then it sends out a command 'run setup'. Once everyone says 'I ran setup' then it says 'run test' and same thing with 'run teardown'. It will keep going until the tests are done. All of this logic btw is defined in test/TestServer/UnitTestFramework.js.

The problem is that we have had failures in tear down that don't seem to be noticed by CI. Looking [here](https://github.com/thaliproject/Thali_CordovaPlugin/blame/vNext_yarong_780/test/TestServer/UnitTestFramework.js#L180) it seems like we should notice when a test fails. And when I look [here](https://github.com/thaliproject/Thali_CordovaPlugin/blob/a9475e1ee5dfc8949748c8216b1ab26427ff2b50/test/www/jxcore/lib/thaliTape.js#L131) it looks like the client is sending information on setup and teardown success or failure. So what's going on?

Did the bug get fixed and we didn't notice?

So the immediate work item is to add tests to testUnitTestFramework that confirm if we throw an exception or if we fail a test (e.g. call t.fail() for example inside of setup or teardown) then the unit test framework will properly notice this.

Then you need to setup a local test of the coordinator (e.g. the code in test/TestServer) to make sure that if we get an error in setup or teardown that this will be properly reflected in the test results.

**Comment by artemjackson** [comment]
@yaronyg I've added a [test file](https://github.com/thaliproject/Thali_CordovaPlugin/pull/836/commits/f0af70021ee2b5e98bc90b9e741cf7259579af1a) that fires test failures during the `setup` and `teardown`.

According to manually running coordinated tests logs all failures were properly handled by `UnitTestFramework`.

So, there are some questions.
1. Should I add an unit-test that checks proper flow of error handling in the `UnitTestFramework`? 
2. If so, there should be an ability to inject a [`testServer`](https://github.com/thaliproject/Thali_CordovaPlugin/blob/vNext_artemjackson_691/test/www/jxcore/lib/thaliTape.js#L208) instead of hard-coding it. With injection of the `testServer` it will be possible to mock it for testing proper error handling of the `UnitTestFramework`.

**Comment by yaronyg** [comment]
We really need to write a test that validates that we properly detect failures in setup and teardown but it's not high enough priority at the moment given that manual testing shows that we handle it properly.
